File: src/common/crypto.c
/*
 * crypto.c — AES-256-GCM payload encryption for DNS tunnel.
 *
 * Uses OpenSSL EVP API. Key derived from PSK via HKDF-SHA256.
 * All encryption uses random nonces (safe for up to ~2^48 messages).
 */
#include "crypto.h"

#include <string.h>
#include <openssl/evp.h>
#include <openssl/rand.h>
#include <openssl/hmac.h>
#include <openssl/sha.h>

/* ── State ────────────────────────────────────────────────────────────────── */

static uint8_t g_key[32];      /* AES-256 key derived from PSK */
static int     g_enabled = 0;  /* 1 if PSK was set */

/* ── HKDF-SHA256 (RFC 5869) ──────────────────────────────────────────────── */

/*
 * HKDF-Extract: PRK = HMAC-SHA256(salt, ikm)
 * HKDF-Expand:  OKM = HMAC-SHA256(PRK, info || 0x01)  [first 32 bytes only]
 *
 * This produces exactly 32 bytes — perfect for AES-256.
 * Implemented manually for maximum OpenSSL version compatibility.
 */
static int hkdf_sha256(const uint8_t *salt, size_t salt_len,
                       const uint8_t *ikm,  size_t ikm_len,
                       const uint8_t *info, size_t info_len,
                       uint8_t *okm, size_t okm_len)
{
    if (okm_len > 32) return -1;  /* single HMAC output block */

    /* Extract: PRK = HMAC-SHA256(salt, IKM) */
    uint8_t prk[32];
    unsigned int prk_len = 32;
    if (!HMAC(EVP_sha256(), salt, (int)salt_len,
              ikm, ikm_len, prk, &prk_len))
        return -1;

    /* Expand: T(1) = HMAC-SHA256(PRK, info || 0x01) */
    uint8_t expand_input[256];
    if (info_len + 1 > sizeof(expand_input)) return -1;
    memcpy(expand_input, info, info_len);
    expand_input[info_len] = 0x01;

    uint8_t t1[32];
    unsigned int t1_len = 32;
    if (!HMAC(EVP_sha256(), prk, 32,
              expand_input, info_len + 1, t1, &t1_len))
        return -1;

    memcpy(okm, t1, okm_len);
    return 0;
}

/* ── Public API ──────────────────────────────────────────────────────────── */

int tunnel_crypto_init(const char *psk)
{
    g_enabled = 0;
    memset(g_key, 0, sizeof(g_key));

    if (!psk || !psk[0])
        return 0;  /* No PSK = encryption disabled, not an error */

    static const uint8_t salt[] = "dns-claw-v1";
    static const uint8_t info[] = "tunnel-key";

    if (hkdf_sha256(salt, sizeof(salt) - 1,
                    (const uint8_t *)psk, strlen(psk),
                    info, sizeof(info) - 1,
                    g_key, sizeof(g_key)) < 0)
        return -1;

    g_enabled = 1;
    return 0;
}

int tunnel_crypto_enabled(void)
{
    return g_enabled;
}

int tunnel_encrypt(const uint8_t *in, size_t in_len,
                   uint8_t *out, size_t *out_len)
{
    if (!g_enabled) return -1;

    /* Layout: [magic:2][nonce:12][ciphertext:in_len][tag:16] */
    uint8_t nonce[CRYPTO_NONCE_LEN];
    if (RAND_bytes(nonce, CRYPTO_NONCE_LEN) != 1)
        return -1;

    /* Write magic header */
    out[0] = CRYPTO_MAGIC_0;
    out[1] = CRYPTO_MAGIC_1;

    /* Write nonce */
    memcpy(out + 2, nonce, CRYPTO_NONCE_LEN);

    /* Encrypt */
    EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
    if (!ctx) return -1;

    int rc = -1;
    int len = 0;
    int ciphertext_len = 0;

    if (EVP_EncryptInit_ex(ctx, EVP_aes_256_gcm(), NULL, NULL, NULL) != 1)
        goto cleanup;
    if (EVP_EncryptInit_ex(ctx, NULL, NULL, g_key, nonce) != 1)
        goto cleanup;

    /* Encrypt plaintext → ciphertext at offset 14 (2 magic + 12 nonce) */
    if (EVP_EncryptUpdate(ctx, out + 2 + CRYPTO_NONCE_LEN, &len,
                          in, (int)in_len) != 1)
        goto cleanup;
    ciphertext_len = len;

    if (EVP_EncryptFinal_ex(ctx, out + 2 + CRYPTO_NONCE_LEN + ciphertext_len,
                            &len) != 1)
        goto cleanup;
    ciphertext_len += len;

    /* Append GCM tag */
    if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_GET_TAG, CRYPTO_TAG_LEN,
                            out + 2 + CRYPTO_NONCE_LEN + ciphertext_len) != 1)
        goto cleanup;

    *out_len = 2 + CRYPTO_NONCE_LEN + (size_t)ciphertext_len + CRYPTO_TAG_LEN;
    rc = 0;

cleanup:
    EVP_CIPHER_CTX_free(ctx);
    return rc;
}

int tunnel_decrypt(const uint8_t *in, size_t in_len,
                   uint8_t *out, size_t *out_len)
{
    if (!g_enabled) return -1;

    /* Minimum size: 2 (magic) + 12 (nonce) + 0 (ciphertext) + 16 (tag) */
    if (in_len < CRYPTO_OVERHEAD) return -1;

    /* Verify magic header */
    if (in[0] != CRYPTO_MAGIC_0 || in[1] != CRYPTO_MAGIC_1)
        return -1;

    const uint8_t *nonce      = in + 2;
    const uint8_t *ciphertext = in + 2 + CRYPTO_NONCE_LEN;
    size_t ct_len = in_len - CRYPTO_OVERHEAD;
    const uint8_t *tag = in + in_len - CRYPTO_TAG_LEN;

    EVP_CIPHER_CTX *ctx = EVP_CIPHER_CTX_new();
    if (!ctx) return -1;

    int rc = -1;
    int len = 0;
    int plaintext_len = 0;

    if (EVP_DecryptInit_ex(ctx, EVP_aes_256_gcm(), NULL, NULL, NULL) != 1)
        goto cleanup;
    if (EVP_DecryptInit_ex(ctx, NULL, NULL, g_key, nonce) != 1)
        goto cleanup;

    /* Decrypt ciphertext → plaintext */
    if (EVP_DecryptUpdate(ctx, out, &len,
                          ciphertext, (int)ct_len) != 1)
        goto cleanup;
    plaintext_len = len;

    /* Set expected GCM tag for verification */
    if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_TAG, CRYPTO_TAG_LEN,
                            (void *)tag) != 1)
        goto cleanup;

    /* Finalize — this verifies the tag (returns 0 if tampered) */
    if (EVP_DecryptFinal_ex(ctx, out + plaintext_len, &len) != 1)
        goto cleanup;  /* Authentication failed — wrong key or tampered */
    plaintext_len += len;

    *out_len = (size_t)plaintext_len;
    rc = 0;

cleanup:
    EVP_CIPHER_CTX_free(ctx);
    return rc;
}

File: web/lib/crypto.ts
/**
 * crypto.ts — AES-256-GCM payload encryption for DNS tunnel.
 *
 * Mirrors src/common/crypto.c. Key derived from PSK via HKDF-SHA256.
 * Wire format: [0xCE, 0x01][nonce:12][ciphertext][tag:16]
 */

import { createHmac, createCipheriv, createDecipheriv, randomBytes } from "crypto";

const MAGIC_0 = 0xce;
const MAGIC_1 = 0x01;
const NONCE_LEN = 12;
const TAG_LEN = 16;
const OVERHEAD = 2 + NONCE_LEN + TAG_LEN; // 30 bytes

/**
 * HKDF-SHA256 (RFC 5869) — extract + expand to produce exactly 32 bytes.
 * Matches the C implementation in crypto.c.
 */
function hkdfSha256(
  salt: Buffer,
  ikm: Buffer,
  info: Buffer,
): Buffer {
  // Extract: PRK = HMAC-SHA256(salt, IKM)
  const prk = createHmac("sha256", salt).update(ikm).digest();

  // Expand: T(1) = HMAC-SHA256(PRK, info || 0x01)
  const expandInput = Buffer.concat([info, Buffer.from([0x01])]);
  const t1 = createHmac("sha256", prk).update(expandInput).digest();

  return t1; // 32 bytes — exactly what AES-256 needs
}

/** Derive AES-256 key from PSK. Returns null if no PSK. */
export function deriveKey(psk: string | undefined): Buffer | null {
  if (!psk || psk.length === 0) return null;

  return hkdfSha256(
    Buffer.from("dns-claw-v1", "utf-8"),
    Buffer.from(psk, "utf-8"),
    Buffer.from("tunnel-key", "utf-8"),
  );
}

/** Encrypt plaintext with AES-256-GCM. Returns [magic][nonce][ct][tag]. */
export function encrypt(plaintext: Buffer, key: Buffer): Buffer {
  const nonce = randomBytes(NONCE_LEN);
  const cipher = createCipheriv("aes-256-gcm", key, nonce);

  const ciphertext = Buffer.concat([cipher.update(plaintext), cipher.final()]);
  const tag = cipher.getAuthTag();

  return Buffer.concat([
    Buffer.from([MAGIC_0, MAGIC_1]),
    nonce,
    ciphertext,
    tag,
  ]);
}

/** Decrypt [magic][nonce][ct][tag] with AES-256-GCM. */
export function decrypt(data: Buffer, key: Buffer): Buffer {
  if (data.length < OVERHEAD) {
    throw new Error("Encrypted data too short");
  }
  if (data[0] !== MAGIC_0 || data[1] !== MAGIC_1) {
    throw new Error("Invalid magic bytes — not an encrypted DNS-CLAW payload");
  }

  const nonce = data.subarray(2, 2 + NONCE_LEN);
  const ciphertext = data.subarray(2 + NONCE_LEN, data.length - TAG_LEN);
  const tag = data.subarray(data.length - TAG_LEN);

  const decipher = createDecipheriv("aes-256-gcm", key, nonce);
  decipher.setAuthTag(tag);

  return Buffer.concat([decipher.update(ciphertext), decipher.final()]);
}

export { OVERHEAD as CRYPTO_OVERHEAD };
